File: android/lib/src/main/kotlin/com/mia21/services/DynamicPromptService.kt
/**
 * DynamicPromptService.kt
 * Mia21
 *
 * Created on January 21, 2026.
 * Copyright © 2026 Mia21. All rights reserved.
 *
 * Description:
 * Service layer for dynamic prompting via OpenAI-compatible endpoint.
 * Allows runtime AI configuration without pre-configuration.
 */

package com.mia21.services

import com.mia21.models.*
import com.mia21.network.APIClient
import com.mia21.network.APIEndpoint
import com.mia21.network.HTTPMethod
import com.mia21.utils.Logger
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.channels.awaitClose
import kotlinx.serialization.json.Json

/**
 * Service for dynamic prompting operations
 */
class DynamicPromptService(
    private val apiClient: APIClient,
    private val defaultUserId: String
) {
    
    private val json = Json { ignoreUnknownKeys = true }
    
    /**
     * Send a completion request using the OpenAI-compatible endpoint
     * @param messages Array of messages including system prompt
     * @param options Configuration options for the request
     * @return The completion response
     */
    suspend fun complete(
        messages: List<ChatMessage>,
        options: DynamicPromptOptions
    ): DynamicPromptResponse {
        Logger.info("Dynamic prompt completion with model: ${options.model}")
        
        val messagesArray = messages.map { msg ->
            mapOf("role" to msg.role.name.lowercase(), "content" to msg.content)
        }
        
        val body = mutableMapOf<String, Any?>(
            "model" to options.model,
            "messages" to messagesArray,
            "stream" to false
        )
        
        options.temperature?.let { body["temperature"] = it }
        options.maxTokens?.let { body["max_tokens"] = it }
        
        // Build custom headers for user/space context
        val headers = mutableMapOf<String, String>()
        options.spaceId?.let { headers["X-Space-Id"] = it }
        headers["X-User-Id"] = options.userId ?: defaultUserId
        
        val endpoint = APIEndpoint(
            path = "/v1/chat/completions",
            method = HTTPMethod.POST,
            body = body,
            headers = headers
        )
        
        val jsonResponse = apiClient.performRequest(endpoint, String::class.java)
        val response = json.decodeFromString<DynamicPromptResponse>(jsonResponse)
        
        Logger.info("Dynamic prompt completed. Model: ${response.model}")
        response.usage?.let { usage ->
            Logger.debug("  Tokens - Prompt: ${usage.promptTokens}, Completion: ${usage.completionTokens}, Total: ${usage.totalTokens}")
        }
        
        return response
    }
    
    /**
     * Stream a completion request using the OpenAI-compatible endpoint
     * @param messages Array of messages including system prompt
     * @param options Configuration options for the request
     * @return Flow of text chunks
     */
    fun streamComplete(
        messages: List<ChatMessage>,
        options: DynamicPromptOptions
    ): Flow<String> = callbackFlow {
        Logger.info("Streaming dynamic prompt with model: ${options.model}")
        
        val messagesArray = messages.map { msg ->
            mapOf("role" to msg.role.name.lowercase(), "content" to msg.content)
        }
        
        val body = mutableMapOf<String, Any?>(
            "model" to options.model,
            "messages" to messagesArray,
            "stream" to true
        )
        
        options.temperature?.let { body["temperature"] = it }
        options.maxTokens?.let { body["max_tokens"] = it }
        
        // Build custom headers for user/space context
        val headers = mutableMapOf<String, String>()
        options.spaceId?.let { headers["X-Space-Id"] = it }
        headers["X-User-Id"] = options.userId ?: defaultUserId
        
        val endpoint = APIEndpoint(
            path = "/v1/chat/completions",
            method = HTTPMethod.POST,
            body = body,
            headers = headers
        )
        
        val job = launch {
            try {
                apiClient.performStreamRequest(endpoint).collect { data ->
                    processStreamData(data)?.let { content ->
                        trySend(content)
                    }
                }
                Logger.info("Dynamic prompt stream completed")
            } catch (e: Exception) {
                Logger.error("Stream failed: ${e.message}")
                close(e)
            }
        }
        
        awaitClose {
            job.cancel()
        }
    }
    
    /**
     * Process stream data and extract text content
     */
    private fun processStreamData(data: String): String? {
        val trimmed = data.trim()
        
        if (trimmed.isEmpty() || trimmed == "[DONE]") {
            return null
        }
        
        return try {
            val chunk = json.decodeFromString<DynamicPromptStreamChunk>(trimmed)
            chunk.choices.firstOrNull()?.delta?.content?.takeIf { it.isNotEmpty() }
        } catch (e: Exception) {
            Logger.debug("Failed to parse stream chunk: ${e.message}")
            null
        }
    }
}

File: ios/Sources/Mia21/Services/DynamicPromptService.swift
//
//  DynamicPromptService.swift
//  Mia21
//
//  Created on January 21, 2026.
//  Copyright © 2026 Mia21. All rights reserved.
//
//  Description:
//  Service layer for dynamic prompting via OpenAI-compatible endpoint.
//  Allows runtime AI configuration without pre-configuration.
//

import Foundation

// MARK: - Dynamic Prompt Service Protocol

protocol DynamicPromptServiceProtocol {
  func complete(
    messages: [ChatMessage],
    options: DynamicPromptOptions
  ) async throws -> DynamicPromptResponse
  
  func streamComplete(
    messages: [ChatMessage],
    options: DynamicPromptOptions,
    onChunk: @escaping (String) -> Void
  ) async throws
}

// MARK: - Dynamic Prompt Service Implementation

final class DynamicPromptService: DynamicPromptServiceProtocol {
  
  // MARK: - Properties
  
  private let apiClient: APIClientProtocol
  private let defaultUserId: String
  
  // MARK: - Initialization
  
  init(apiClient: APIClientProtocol, defaultUserId: String) {
    self.apiClient = apiClient
    self.defaultUserId = defaultUserId
  }
  
  // MARK: - Public Methods
  
  /// Send a completion request using the OpenAI-compatible endpoint
  /// - Parameters:
  ///   - messages: Array of messages including system prompt
  ///   - options: Configuration options for the request
  /// - Returns: The completion response
  func complete(
    messages: [ChatMessage],
    options: DynamicPromptOptions
  ) async throws -> DynamicPromptResponse {
    logInfo("Dynamic prompt completion with model: \(options.model)")
    
    // Build messages array for API
    let messagesArray = messages.map { msg -> [String: String] in
      ["role": msg.role.rawValue, "content": msg.content]
    }
    
    var body: [String: Any] = [
      "model": options.model,
      "messages": messagesArray,
      "stream": false
    ]
    
    if let temperature = options.temperature {
      body["temperature"] = temperature
    }
    if let maxTokens = options.maxTokens {
      body["max_tokens"] = maxTokens
    }
    
    // Build custom headers for user/space context
    var headers: [String: String] = [:]
    if let spaceId = options.spaceId {
      headers["X-Space-Id"] = spaceId
    }
    let userId = options.userId ?? defaultUserId
    headers["X-User-Id"] = userId
    
    let endpoint = APIEndpoint(
      path: "/v1/chat/completions",
      method: .post,
      body: body,
      headers: headers
    )
    
    let response: DynamicPromptResponse = try await apiClient.performRequest(endpoint)
    
    logInfo("Dynamic prompt completed. Model: \(response.model)")
    if let usage = response.usage {
      logDebug("  Tokens - Prompt: \(usage.promptTokens), Completion: \(usage.completionTokens), Total: \(usage.totalTokens)")
    }
    
    return response
  }
  
  /// Stream a completion request using the OpenAI-compatible endpoint
  /// - Parameters:
  ///   - messages: Array of messages including system prompt
  ///   - options: Configuration options for the request
  ///   - onChunk: Callback invoked for each text chunk
  func streamComplete(
    messages: [ChatMessage],
    options: DynamicPromptOptions,
    onChunk: @escaping (String) -> Void
  ) async throws {
    logInfo("Streaming dynamic prompt with model: \(options.model)")
    
    // Build messages array for API
    let messagesArray = messages.map { msg -> [String: String] in
      ["role": msg.role.rawValue, "content": msg.content]
    }
    
    var body: [String: Any] = [
      "model": options.model,
      "messages": messagesArray,
      "stream": true
    ]
    
    if let temperature = options.temperature {
      body["temperature"] = temperature
    }
    if let maxTokens = options.maxTokens {
      body["max_tokens"] = maxTokens
    }
    
    // Build custom headers for user/space context
    var headers: [String: String] = [:]
    if let spaceId = options.spaceId {
      headers["X-Space-Id"] = spaceId
    }
    let userId = options.userId ?? defaultUserId
    headers["X-User-Id"] = userId
    
    let endpoint = APIEndpoint(
      path: "/v1/chat/completions",
      method: .post,
      body: body,
      headers: headers
    )
    
    let stream = try await apiClient.performStreamRequest(endpoint)
    
    for try await data in stream {
      if let line = String(data: data, encoding: .utf8) {
        processStreamLine(line, onChunk: onChunk)
      }
    }
    
    logInfo("Dynamic prompt stream completed")
  }
  
  // MARK: - Private Methods
  
  private func processStreamLine(_ line: String, onChunk: @escaping (String) -> Void) {
    let trimmed = line.trimmingCharacters(in: .whitespaces)
    
    // Skip empty lines
    if trimmed.isEmpty {
      return
    }
    
    // Check for SSE data prefix
    if line.hasPrefix("data: ") {
      let dataContent = String(line.dropFirst(6))
      
      // Check for stream end marker
      if dataContent == "[DONE]" {
        return
      }
      
      // Parse the JSON chunk
      if let data = dataContent.data(using: .utf8),
         let chunk = try? JSONDecoder().decode(DynamicPromptStreamChunk.self, from: data) {
        // Extract content from the delta
        if let content = chunk.choices.first?.delta.content, !content.isEmpty {
          onChunk(content)
        }
      }
    }
  }
}

